FAI-437: CF: TrustyService endpoint for UI to initiate CF request

manstis opened the pull request:
See:-
- https://issues.redhat.com/browse/FAI-437
- https://issues.redhat.com/browse/FAI-438
- https://issues.redhat.com/browse/FAI-440

This PR covers:
- Request from the UI
- Storage of CF request
- Sending Event to `ExplainabilityService` for CF calculation.

This PR does not cover:
- Changes to `ExplainabilityService` to handle different explainability requests. `ExplanationServiceImpl` only has LIME's `LocalExplainer` injected and hence no lookup/indirection is currently performed for different types of request.
- Changes to `ExplainabilityService` REST handler (i.e. only LIME is exposed). Is CF to be exposed via REST? It would appear that whatever changes are made to `ExplanationServiceImpl` will also cover expose via REST.
- Receipt and storage of "results" from the CF explainer.
- Retrieval of "results" by the UI.

More Unit/Integration Tests are required, but I wanted to gather opinions.

I am inclined to propose merge of this PR once the "request" path is complete (with tests).

The "results" path to be completed as a separate PR.

It is becoming a monster.

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

manstis commented on explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/BaseExplainabilityRequestDto.java:
```diff
@@ -48,10 +59,10 @@
     @JsonProperty("outputs")
     private Map<String, TypedValue> outputs;
 
-    private ExplainabilityRequestDto() {
```
The old (LIME) explainability request/result has been refactored to support LIME and CF (and others).

manstis commented on explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualExplainabilityResultDto.java:
```diff
@@ -0,0 +1,37 @@
+ */
+package org.kie.kogito.explainability.api;
+
+public class CounterfactualExplainabilityResultDto extends BaseExplainabilityResultDto {
```
Pretty empty at the moment.. I need to work on the "result" path in collaboration with @ruivieira.

ruivieira replied:
@manstis ack.

manstis commented on trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/TrustyServiceImpl.java:
```diff
@@ -180,10 +186,91 @@ public DMNModelWithMetadata getModelById(ModelIdentifier modelIdentifier) {
+            List<CounterfactualSearchDomain> searchDomains) {
+        Decision decision = getDecisionById(executionId);
+
+        Map<String, TypedValue> originalInputs = decision.getInputs() != null
```
Here I map UI<->Events models.

manstis commented on trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/messaging/incoming/ExplainabilityResultConsumer.java:
```diff
@@ -62,20 +64,26 @@ public ExplainabilityResultConsumer(TrustyService service, ObjectMapper mapper)
-        Map<String, String> outcomeNameToIdMap = decision == null
-                ? Collections.emptyMap()
-                : decision.getOutcomes().stream().collect(Collectors.toUnmodifiableMap(DecisionOutcome::getOutcomeName, DecisionOutcome::getOutcomeId));
+        if (dto instanceof LIMEExplainabilityResultDto) {
```
I need to add support for the CF result... for not it'll barf.

manstis commented on trusty/trusty-service/trusty-service-common/src/test/java/org/kie/kogito/trusty/service/common/api/ExplainabilityApiV1IT.java:
```diff
@@ -115,6 +144,74 @@ void testSalienciesWithoutExplainabilityResult() {
+                .filter(new ResponseLoggingFilter())
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(getCounterfactualJsonRequest())
+                .when().post("/executions/decisions/" + TEST_EXECUTION_ID + "/explanations/counterfactuals")
```
I moved it to a file @r00ta :-)

manstis commented on trusty/trusty-service/trusty-service-common/src/test/java/org/kie/kogito/trusty/service/common/api/ExplainabilityApiV1IT.java:
```diff
@@ -115,6 +144,74 @@ void testSalienciesWithoutExplainabilityResult() {
+                .filter(new RequestLoggingFilter())
+                .filter(new ResponseLoggingFilter())
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(getCounterfactualJsonRequest())
```
I moved it to a file @r00ta :-)

manstis commented on explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java:
```diff
@@ -72,8 +73,8 @@ public ExplanationServiceImpl(LocalExplainer<Map<String, Saliency>> localExplain
-    private static ExplainabilityResultDto createSucceededResultDto(String executionId, Map<String, Saliency> saliencies) {
-        return ExplainabilityResultDto.buildSucceeded(
+    private static BaseExplainabilityResultDto createSucceededResultDto(String executionId, Map<String, Saliency> saliencies) {
+        return LIMEExplainabilityResultDto.buildSucceeded(
```
Everything defaults to LIME at the moment.

IDK who is to make the changes to the request/service/predicator/result mechanism?

@ruivieira would this be in your area of responsibility?

manstis replied:
I have added a commit that throws an `IllegalArgumentExeption` if a Couterfactual request is received.

r00ta commented:
Hi @manstis , thank you very much I'll review this tomorrow :)

r00ta commented on trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/TrustyServiceImpl.java:
```diff
@@ -180,10 +186,91 @@ public DMNModelWithMetadata getModelById(ModelIdentifier modelIdentifier) {
+                        .collect(HashMap::new, (m, v) -> m.put(v.getName(), MessagingUtils.modelToTracingTypedValue(v)), HashMap::putAll)
+                : Collections.emptyMap();
+
+        Map<String, CounterfactualSearchDomainDto> searchDomainDtos = searchDomains != null
+                ? searchDomains.stream()
+                        .collect(HashMap::new, (m, v) -> m.put(v.getName(), MessagingUtils.modelToCounterfactualSearchDomainDto(v)), HashMap::putAll)
+                : Collections.emptyMap();
+
```
I might be missing something, but why is this conversion needed? 
I mean: this call is converting the `CounterfactualSearchDomain` obj to the `CounterfactualSearchDomainDto` obj, but actually they contain the same amount of information right (they just diff in the structure)? I'm wondering if we could keep the same structure from UI <-> trusty <-> explainability so that we avoid having one DTO for the UI request and one DTO for the explainability request message. wdyt?

manstis replied:
Marking as resolved to keep the open issues clear.

r00ta reviewed (commented):
Hey @manstis , I have a question related to the fact that we use one DTO for the CF request from the UI and another one for the explainability request message. I might be missing something, do you mind to clarify it to me before I move on with the review? 
Thank you very much :+1:

manstis commented:
Hi @r00ta I'll answer all your questions relating to DTOs and mapping here.

> Hey @manstis , I have a question related to the fact that we use one DTO for the CF request from the UI and another one for the explainability request message. I might be missing something, do you mind to clarify it to me before I move on with the review?

The existing (LIME) explainability messages use their own DTOs that I have re-purposed as much as possible for CF too (the change to `BaseExplainabilityRequestDto` and `BaseExplainabilityResultDto` now sub-classed for LIME and CF). The CF sub-class of `BaseExplainabilityRequestDto` needs an additional field for "search domains" which is the work you have seen for `CounterfactualSearchDomainDto`. This keeps the whole `CounterfactualExplainabilityRequestDto` consistent with the `LIMEExplainabilityRequestDto` (i.e. different structure of `TypedValue`/`TypedVariable`). Not ideal, as we know from this discussion we are having about mapping DTOs at service boundaries, but consistent with what already exists.

If I was crazy enough to undertake a refactor of all DTOs across all services in this PR I'd end up refactoring some of the explainability service itself which is really not something I wanted to entertain right now. So, I've opted for consistency vs purity.

r00ta commented:
> Hi @r00ta I'll answer all your questions relating to DTOs and mapping here.
> 
> > Hey @manstis , I have a question related to the fact that we use one DTO for the CF request from the UI and another one for the explainability request message. I might be missing something, do you mind to clarify it to me before I move on with the review?
> 
> The existing (LIME) explainability messages use their own DTOs that I have re-purposed as much as possible for CF too (the change to `BaseExplainabilityRequestDto` and `BaseExplainabilityResultDto` now sub-classed for LIME and CF). The CF sub-class of `BaseExplainabilityRequestDto` needs an additional field for "search domains" which is the work you have seen for `CounterfactualSearchDomainDto`. This keeps the whole `CounterfactualExplainabilityRequestDto` consistent with the `LIMEExplainabilityRequestDto` (i.e. different structure of `TypedValue`/`TypedVariable`). Not ideal, as we know from this discussion we are having about mapping DTOs at service boundaries, but consistent with what already exists.
> 
> If I was crazy enough to undertake a refactor of all DTOs across all services in this PR I'd end up refactoring some of the explainability service itself which is really not something I wanted to entertain right now. So, I've opted for consistency vs purity.
> 
> So, right now, we have one set of DTOs for UI<->`TrustyService`<->`TrustyStorage` and another set of DTOs for Cloud Events<->`ExplainabilityService`. It would be great to have a new module somewhere for all DTOs used in all of these services/boundaries.

Hey @manstis , the refactoring of the DTOs is absolutely out of scope from this PR. I see that you kept consistency with the current implementation, I think it is the right strategy and you did it great! 

I see now that `CounterfactualSearchDomainDto` is following the same logic of the `TypedValue`.. and if you were using `CounterfactualSearchDomain` instead you would have introduced the `TypedVariable` as well in the same cloud event. 
I think the root of all the pain is that we two different implementation (`TypedValue` and `TypedVariable`) to represent the same thing, and we use them here and there with a lot of conversion and no clear benefits. Is that correct @manstis ?

manstis commented:
Hi @r00ta

> Hey @manstis , the refactoring of the DTOs is absolutely out of scope from this PR as you pointed out. I see that you kept consistency with the current implementation, I think it is the right strategy and you did it great!
> 
> I see now that `CounterfactualSearchDomainDto` is following the same logic of the `TypedValue`.. and if you were using `CounterfactualSearchDomain` instead you would have introduced the `TypedVariable` as well in the same cloud event.
> Is that correct @manstis ?
> 
> I think the root of all the pain is that we two different implementation (`TypedValue` and `TypedVariable`) to represent the same thing, and we use them here and there with a lot of conversion and no clear benefits.

Correct on all accounts. When this CF work is complete it'd be great to have a single module containing single DTOs that can be used for everything: Tracing Addon, Cloud Events, Trusty Service, Trusty Storage and Explainability Service; killing the mapping craziness.

r00ta commented on trusty/trusty-service/trusty-service-common/src/test/resources/requests/counterfactualRequest.json:
```diff
@@ -0,0 +1,44 @@
+      "name": "age",
+      "typeRef": "number"
+    },
+    {
+      "isFixed": false,
+      "name": "income",
+      "typeRef": "number",
+      "domain": {
+        "type": "numerical",
+        "lowerBound": 0,
+        "upperBound": 1000
+      }
+    },
```
How does it work with structured data? 
for example
```
{
  "Loan": {
    "Required": <numerical constraint lb= 100, up= 1000>
    "Installments": <numerical constraint lb= 10, up= 100>
  }
}
```
Is the current structure capable to handle this case? The comment could be on the java class, but looking at this data it's easier to see the point

manstis replied:
See https://github.com/kiegroup/kogito-apps/pull/747/commits/1f97651045a224025ca6589fbc014cba7df68eb9

r00ta replied:
Super! Thank you :)

r00ta reviewed (commented):
Hey @manstis , I still miss half of the changes but I start asking a couple of questions in the meanwhile! (will continue with the review later)

manstis commented on explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualSearchDomainDto.java:
```diff
@@ -0,0 +1,101 @@
+        return kind == CounterfactualSearchDomainDto.Kind.UNIT;
+    }
+
+    public CounterfactualSearchDomainUnitDto toUnit() {
+        if (!isUnit()) {
+            throw new IllegalStateException(String.format("Can't convert TypedValue of kind %s to UNIT", kind));
+        }
+        return (CounterfactualSearchDomainUnitDto) this;
+    }
```
No, not used.. but IDK if it's required by @ruivieira at some point (it exists on the `TypedValue` class too).

manstis commented on explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualSearchDomainDto.java:
```diff
@@ -0,0 +1,101 @@
+        return kind == CounterfactualSearchDomainDto.Kind.STRUCTURE;
+    }
+
+    public CounterfactualSearchDomainStructureDto toStructure() {
+        if (!isStructure()) {
+            throw new IllegalStateException(String.format("Can't convert TypedValue of kind %s to STRUCTURE", kind));
+        }
+        return (CounterfactualSearchDomainStructureDto) this;
+    }
```
Likewise; re: usage.

r00ta commented on trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/Counterfactual.java:
```diff
@@ -0,0 +1,97 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class Counterfactual {
```
`CounterfactualRequest`?

manstis replied:
Renamed in new commit along with some other related things.

r00ta commented on trusty/trusty-storage/trusty-storage-common/src/main/java/org/kie/kogito/trusty/storage/common/TrustyStorageServiceImpl.java:
```diff
@@ -37,12 +39,22 @@
     }
+
+    @Override
+    public Storage<String, Counterfactual> getCounterfactualStorage() {
```
`getCounterfactualRequestStorage`?

manstis replied:
Renamed in new commit along with some other related things.

r00ta reviewed (approved):
Thank you very much @manstis !! I think the changes are clear and the goal of the PR is achieved very well. Unfortunately we have an explosion of DTO/converters as well as marshallers for the infinispan implementation that we should not solve in this PR. 
My suggestion is to create the following tickets
1) refactoring all the DTOs to reduce the amount of conversions
2) refactoring the data structures that we store in infinispan to have something similar to what we have in redis: we can create a custom object where we index only the properties that we need (for example the `executionId` and the `counterfactualId`), and the object is simply serialized to a string. The marshaller will then take the entire string and deserialize it. With this approach, we are going to improve the infinispan performances and we can remove basically all the marshallers for the nested properties (for example the `CounterfactualDomainCategoricalMarshaller`, `CounterfactualDomainMarshaller`, `CounterfactualDomainNumericalMarshaller`, `CounterfactualSearchDomainMarshaller`, `MessageMarshaller`, `TypedVariableWithValueMarshaller` and so on).

Of course this is just a proposal for the team, let me know what do you think about those points ^^ @danielezonca @manstis .

manstis commented on explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java:
```diff
@@ -56,7 +57,7 @@ public ExplanationServiceImpl(LocalExplainer<Map<String, Saliency>> localExplain
 
     @Override
-    public CompletionStage<ExplainabilityResultDto> explainAsync(
+    public CompletionStage<BaseExplainabilityResultDto> explainAsync(
```
Done in new commit. I've done it at the point where we convert an incoming Explainability request to the internal form used by the Explainability Service (what!?!?! More DTO conversion!?!!)

manstis commented on trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/messaging/incoming/TraceEventConsumer.java:
```diff
@@ -62,7 +62,10 @@ protected void internalHandleCloudEvent(CloudEvent cloudEvent, TraceEvent payloa
 
         if (traceEventType == TraceEventType.DMN) {
-            service.processDecision(cloudEvent.getId(), payload.getHeader().getResourceId().getServiceUrl(), TraceEventConverter.toDecision(payload, cloudEvent.getSource().toString()));
+            String sourceUrl = cloudEvent.getSource().toString();
```
Erm, this class is inside `io.cloudevents.CloudEventAttributes`.

Was you going to submit a PR to them? It's a breaking change across the whole Universe.

manstis commented:
Hi @r00ta @danielezonca I believe the two things identified are correct:

> 1. Refactoring all the DTOs to reduce the amount of conversions
> 2. Refactoring the data structures that we store in infinispan....

@r00ta was you making the tickets or shall I?

ruivieira reviewed (commented):
Hi @manstis, thank you, great work!

LGTM, I'll just write here what we discussed elsewhere:

- It would be worth to have some kind of extra type information for search domains, specifically to distinguish `int` and `double` in numerical bounds
- I'll take the task of writing the explainer producer for CFs (tracked at [FAI-455](https://issues.redhat.com/browse/FAI-455))

manstis commented:
@r00ta I've renamed the classes:

https://github.com/kiegroup/kogito-apps/pull/747/commits/f91c7846bc1b7baae00087b3d303acdd29c071f6

@ruivieira @r00ta I've changed the _numeric_ Search Domain to be a more generic _range_ and now use `JsonNode` for both the upper and lower bounds.

https://github.com/kiegroup/kogito-apps/pull/747/commits/82cea8065ab5e661420acf5a3a592d3106ad3a80

manstis commented:
Jenkins retest this

manstis commented:
Jenkins retest this

r00ta commented:
> 2\. Refactoring the data structures that we store in infinispan

Hi @manstis @danielezonca , I created the following tickets. Feel free to improve the descriptions/add comments :) 

Thank you very much for the nice work @manstis !

https://issues.redhat.com/browse/FAI-458
https://issues.redhat.com/browse/FAI-457

r00ta commented on explainability/explainability-api/src/main/java/org/kie/kogito/explainability/api/CounterfactualDomainCategoricalDto.java:
```diff
@@ -0,0 +1,46 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+
+public class CounterfactualDomainCategoricalDto extends CounterfactualDomainDto {
```
https://issues.redhat.com/browse/FAI-457

ruivieira reviewed (approved):
@manstis LGTM 👍, thanks!

manstis commented:
Jenkins retest this

manstis commented:
@cristianonicolai Do you know why `Kogito Apps :: Integration Tests :: Jobs Service :: Quarkus` may be failing?

manstis commented:
@cristianonicolai Do you know why `Kogito Apps :: Integration Tests :: Jobs Service :: Quarkus` may be failing?

manstis commented:
Jenkins retest this

manstis commented:
Jenkins retest this

manstis commented:
Jenkins retest this

manstis commented:
@danielezonca @jiripetrlik can you please review.

danielezonca commented on trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/messaging/incoming/ExplainabilityResultConsumer.java:
```diff
@@ -62,20 +64,26 @@ public ExplainabilityResultConsumer(TrustyService service, ObjectMapper mapper)
-                        .map(e -> saliencyFrom(outcomeNameToIdMap.get(e.getKey()), e.getKey(), e.getValue()))
-                        .collect(Collectors.toList());
-        return new ExplainabilityResult(dto.getExecutionId(), statusFrom(dto.getStatus()), dto.getStatusDetails(), saliencies);
+        //TODO {manstis} I need to handle different ExplainabilityResults (see CounterfactualResult too).
```
Do you want to add a ticket id/remove this TODO?

manstis replied:
The work is in progress on my PR https://github.com/kiegroup/kogito-apps/pull/752.

It's take a week to get this green I didn't really want to rebuild (with new commit) for the sake of a comment.

danielezonca commented on explainability/explainability-service/src/main/java/org/kie/kogito/explainability/models/ExplainabilityRequest.java:
```diff
@@ -57,12 +58,16 @@ public ModelIdentifier getModelIdentifier() {
+                    dto.getInputs(),
+                    dto.getOutputs());
+        }
+        //TODO ExplanationServiceImpl only supports a LIME LocalExplainer so we need to fail fast for other types.
```
Can you please add a ticket id?

manstis replied:
The work is in progress on my PR https://github.com/kiegroup/kogito-apps/pull/752.

It's take a week to get this green I didn't really want to rebuild (with new commit) for the sake of a comment.

danielezonca replied:
Sure, I understand the pain :)
Fine for me to postpone similar changes to another PR and merge this 👍

danielezonca reviewed (approved):
Thanks for the PR, approving with minor comment. I agree with you and @r00ta comments about the refactor to simplify 👍

manstis commented on trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualSearchDomain.java:
```diff
@@ -0,0 +1,88 @@
+    // Test methods
+    //-------------
+
+    public void setFixed(Boolean isFixed) {
+        this.isFixed = isFixed;
+    }
+
+    public void setDomain(CounterfactualDomain domain) {
+        this.domain = domain;
+    }
```
Sadly the pre-existing test infrastructure will not work with `protected` access modifiers.

I had suggested the `public` setters be replaced with use of reflection.

Anyway, the DTOs and Infinispan marshalling strategy are both up for refactoring.

When that is completed most (if not all) of these marshallers and related unit tests may not exist.

I suggest we live with the tests as they are for this PR.

manstis commented on trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualResult.java:
```diff
@@ -0,0 +1,66 @@
+    // Test methods
+    //-------------
+
+    public void setCounterfactualId(String counterfactualId) {
+        this.counterfactualId = counterfactualId;
+    }
+
+    public void setSolutionId(String solutionId) {
+        this.solutionId = solutionId;
+    }
```
Same as above.

manstis commented on trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualRequest.java:
```diff
@@ -0,0 +1,97 @@
+    //-------------
+    // Test methods
+    //-------------
+    public void setExecutionId(String executionId) {
+        this.executionId = executionId;
+    }
+
+    public void setCounterfactualId(String counterfactualId) {
+        this.counterfactualId = counterfactualId;
+    }
+
+    public void setGoals(Collection<TypedVariableWithValue> goals) {
+        this.goals = goals;
+    }
+
+    public void setSearchDomains(Collection<CounterfactualSearchDomain> searchDomains) {
+        this.searchDomains = searchDomains;
+    }
```
Same as above.

manstis commented on trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainCategorical.java:
```diff
@@ -0,0 +1,60 @@
+    // Test methods
+    //-------------
+
+    public void setCategories(Collection<JsonNode> categories) {
+        this.categories = categories;
+    }
```
Same as above.

manstis commented on trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainRange.java:
```diff
@@ -0,0 +1,72 @@
+    // Test methods
+    //-------------
+
+    public void setLowerBound(JsonNode lowerBound) {
+        this.lowerBound = lowerBound;
+    }
+
+    public void setUpperBound(JsonNode upperBound) {
+        this.upperBound = upperBound;
+    }
```
Same as above.

manstis commented on trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomain.java:
```diff
@@ -0,0 +1,53 @@
+    // Test methods
+    //-------------
+
+    public void setType(Type type) {
+        this.type = type;
+    }
```
Same as above.
